File: merge_sort.py
#merge-sort implementation
def merge(larr,rarr):
    i=0
    j=0
    k=0
    arr=[]
    while(i<len(larr) and j<len(rarr)):
        if larr[i] >rarr[j]:
            arr.append(rarr[j])
            j+=1
            k+=1
        else:
            arr.append(larr[i])
            i+=1
            k+=1
    while(i<len(larr)):
         arr.append(larr[i])
         i+=1
         k+=1
    while(j<len(rarr)):
        arr.append(rarr[j])
        j+=1
        k+=1
    return arr

def merge_sort(arr):
    if len(arr)==1:
        return arr
    
    
    mid=(len(arr)-1)//2
    left=merge_sort(arr[0:mid+1])
    right=merge_sort(arr[mid+1:len(arr)])
    return merge(left,right)


arr=[5,4,3,2,1]
print("Before sorting")
print(arr)
arr=merge_sort(arr)
print(arr)

File: merge_sort.cpp
// Merge sort
#include <bits/stdc++.h>
using namespace std;
void merge(int arr[], int left, int right, int mid)
{

    int lsize = mid - left + 1;
    int rsize = right - mid;
    int larr[lsize];
    int rarr[rsize];
    for (int i = 0; i < lsize; i++)
    {
        larr[i] = arr[left + i];
    }
    for (int j = 0; j < rsize; j++)
    {
        rarr[j] = arr[mid + 1 + j];
    }

    int i = 0;
    int j = 0;
    int k = left;
    while (i < lsize && j < rsize)
    {
        if (larr[i] > rarr[j])
        {
            arr[k] = rarr[j];
            j++;
            k++;
        }
        else
        {
            arr[k] = larr[i];
            i++;
            k++;
        }
    }
    while (i < lsize)
    {
        arr[k] = larr[i];
        i++;
        k++;
    }
    S while (j < rsize)
    {
        arr[k] = rarr[j];
        j++;
        k++;
    }
}
void mergeSort(int arr[], int left, int right)
{
    int mid;
    if (left < right)
    {
        mid = (left + right) / 2;
        mergeSort(arr, left, mid);      // To divide left parts into sub-parts  //It's good to write left instead of 0(Execution Time)
        mergeSort(arr, mid + 1, right); // To divide right parts int sub-parts
        merge(arr, left, right, mid);
    }
}

int main()
{
    int arr[] = {5, 4, 3, 2, 1, 10, 6, 7, 12, 34, 56, 78, 1, 23, 56};
    int size = sizeof(arr) / sizeof(int);
    cout << "Array before sorting" << endl;
    for (int i = 0; i < size; i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
    mergeSort(arr, 0, size - 1);
    cout << "Array after sorting" << endl;
    for (int i = 0; i < size; i++)
    {
        cout << arr[i] << " ";
    }
}